pkg/operators/klusterlet/controllers/addoncontroller/addoncontroller.go

```diff
@@ -0,0 +1,78 @@
+package addoncontroller
+
+import (
+	"context"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	coreinformer "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"open-cluster-management.io/registration-operator/pkg/helpers"
+)
+
+const (
+	imagePullSecret                    = "open-cluster-management-image-pull-credentials"
+	addonInstallNamespaceAnnotationKey = "addon.open-cluster-management.io/namespace"
+)
+
+// AddonController is used to sync pull image secret from operator namespace to addon namespaces(with annotation "addon.open-cluster-management.io/namespace":"true")
+type addonPullImageSecretController struct {
+	operatorNamespace string
+	namespaceInformer coreinformer.NamespaceInformer
+	kubeClient        kubernetes.Interface
+	recorder          events.Recorder
+}
+
+func NewAddonPullImageSecretController(kubeClient kubernetes.Interface, operatorNamespace string, namespaceInformer coreinformer.NamespaceInformer, recorder events.Recorder) factory.Controller {
+	ac := &addonPullImageSecretController{
+		operatorNamespace: operatorNamespace,
+		namespaceInformer: namespaceInformer,
+		kubeClient:        kubeClient,
+		recorder:          recorder,
+	}
+	return factory.New().WithInformersQueueKeyFunc(func(o runtime.Object) string {
+		namespace := o.(*corev1.Namespace)
+		if namespace.Annotations[addonInstallNamespaceAnnotationKey] != "true" {
+			return ""
+		}
+		return namespace.GetName()
+	}, namespaceInformer.Informer()).WithSync(ac.sync).ToController("AddonPullImageSecretController", recorder)
+}
+
+func (c *addonPullImageSecretController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	var err error
+
+	// Sync secret if namespace is created
+	namespace := controllerContext.QueueKey()
+	if namespace == "" {
+		return nil
+	}
+
+	// If namespace does't have addon annotation, return
+	ns, err := c.kubeClient.CoreV1().Namespaces().Get(ctx, namespace, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if ns.Annotations[addonInstallNamespaceAnnotationKey] != "true" {
+		return nil
+	}
+
+	_, _, err = helpers.SyncSecret(
+		ctx,
+		c.kubeClient.CoreV1(),
+		c.kubeClient.CoreV1(),
+		c.recorder,
+		c.operatorNamespace,
+		imagePullSecret,
+		namespace,
+		imagePullSecret,
+		[]metav1.OwnerReference{},
+	)
+	if err != nil {
+		return err
+	}
+	return nil
```

Comment:
do we need to remove the secret when annotation is removed?

Reply:
Currently, we should keep the secret because **users could also create a namespace with the secret**. 
So `(a namespace) + (With image pull secret) + (Without the label)` is a **legal** situation.

Reply:
Agreed to keep secret. @zhiweiyin318 @elgnay